File: framework/CritterStateDrop.cpp
#include "CritterStateDrop.h"

string CritterStateDrop::name = "CritterStateDrop";

CritterStateDrop::CritterStateDrop() {
  power_source = SHORE;
  bus_voltage = 0;
  batv40 = 0;
  batv160 = 0;
  batv280 = 0;
  accel.x = 0;
  accel.y = 0;
  accel.z = 0;
  mag.x = 0;
  mag.y = 0;
  mag.z = 0;
  motor100.velocity = 0;
  motor100.current = 0;
  motor100.temp = 0;
  motor220.velocity = 0;
  motor220.current = 0;
  motor220.temp = 0;
  motor340.velocity = 0;
  motor340.current = 0;
  motor340.temp = 0;
  cycle_time = 0;

}

CritterStateDrop::~CritterStateDrop() {
  
}

int CritterStateDrop::getSize() {
  return sizeof(power_source) +
         sizeof(bus_voltage) +
         sizeof(batv40) + sizeof(batv160) + sizeof(batv280) +
         sizeof(motor100) + sizeof(motor220) + sizeof(motor340) +
         sizeof(accel) + sizeof(mag) + 
         sizeof(rotation) +
         sizeof(ir_distance) +
         sizeof(ir_light) + 
         sizeof(light) + 
         sizeof(thermal) +
         sizeof(bump) +
         sizeof(error_flags) +
         sizeof(cycle_time);
}

void CritterStateDrop::writeArray(void *d) {
  
  char *data = (char *)d;
  memcpy(data, &power_source, sizeof(power_source));
  data += sizeof(power_source);
  
  memcpy(data, &bus_voltage, sizeof(bus_voltage));
  data += sizeof(bus_voltage);
  
  memcpy(data, &batv40, sizeof(batv40));
  data += sizeof(batv40);
  
  memcpy(data, &batv160, sizeof(batv160));
  data += sizeof(batv160);
  
  memcpy(data, &batv280, sizeof(batv280));
  data += sizeof(batv280);
  
  memcpy(data, &motor100, sizeof(motor100));
  data += sizeof(motor100);
  
  memcpy(data, &motor220, sizeof(motor220));
  data += sizeof(motor220);
  
  memcpy(data, &motor340, sizeof(motor340));
  data += sizeof(motor340);
  
  memcpy(data, &accel, sizeof(accel));
  data += sizeof(accel);
  
  memcpy(data, &mag, sizeof(mag));
  data += sizeof(mag);
  
  memcpy(data, &rotation, sizeof(rotation));
  data += sizeof(rotation);
  
  memcpy(data, &ir_distance, sizeof(ir_distance));
  data += sizeof(ir_distance);
  
  memcpy(data, &ir_light, sizeof(ir_light));
  data += sizeof(ir_light);
  
  memcpy(data, &light, sizeof(light));
  data += sizeof(light);
  
  memcpy(data, &thermal, sizeof(thermal));
  data += sizeof(thermal);
  
  memcpy(data, &bump, sizeof(bump));
  data += sizeof(bump);
  
  memcpy(data, &error_flags, sizeof(error_flags));
  data += sizeof(error_flags);
  
  memcpy(data, &cycle_time, sizeof(cycle_time));
  data += sizeof(cycle_time);

}

void CritterStateDrop::readArray(void *d) {

  char *data = (char *)d;
  
  memcpy(&power_source, data, sizeof(power_source));
  data += sizeof(power_source);
  
  memcpy(&bus_voltage, data, sizeof(bus_voltage));
  data += sizeof(bus_voltage);
  
  memcpy(&batv40, data, sizeof(batv40));
  data += sizeof(batv40);
  
  memcpy(&batv160, data, sizeof(batv160));
  data += sizeof(batv160);
  
  memcpy(&batv280, data, sizeof(batv280));
  data += sizeof(batv280);
  
  memcpy(&motor100, data, sizeof(motor100));
  data += sizeof(motor100);
  
  memcpy(&motor220, data, sizeof(motor220));
  data += sizeof(motor220);
  
  memcpy(&motor340, data, sizeof(motor340));
  data += sizeof(motor340);
  
  memcpy(&accel, data, sizeof(accel));
  data += sizeof(accel);
  
  memcpy(&mag, data, sizeof(mag));
  data += sizeof(mag);
  
  memcpy(&rotation, data, sizeof(rotation));
  data += sizeof(rotation);
  
  memcpy(&ir_distance, data, sizeof(ir_distance));
  data += sizeof(ir_distance);
  
  memcpy(&ir_light, data, sizeof(ir_light));
  data += sizeof(ir_light);
  
  memcpy(&light, data, sizeof(light));
  data += sizeof(light);
  
  memcpy(&thermal, data, sizeof(thermal));
  data += sizeof(thermal);
  
  memcpy(&bump, data, sizeof(bump));
  data += sizeof(bump);
  
  memcpy(&error_flags, data, sizeof(error_flags));
  data += sizeof(error_flags);
  
  memcpy(&cycle_time, data, sizeof(cycle_time));
  data += sizeof(cycle_time);

}


string CritterStateDrop::toString(string offset) {

  string value;
  char buf[100];

  value += "\nPower Source: ";
  switch(power_source) {
    case SHORE:
      value += "SHORE\n";
      break;
    case BAT40:
      value += "BAT40\n";
      break;
    case BAT160:
      value += "BAT160\n";
      break;
    case BAT280:
      value += "BAT280\n";
      break;
    default:
      value += "ERROR?\n";
      break;
  }
  sprintf(buf, "System Voltage: %u\n", bus_voltage);
  value += buf;
  sprintf(buf, "Battery 40 Voltage: %u\n", batv40);
  value += buf;
  sprintf(buf, "Battery 160 Voltage: %u\n", batv160);
  value += buf;
  sprintf(buf, "Battery 280 Voltage: %u\n\n", batv280);
  value += buf;
  sprintf(buf, "Motor100: %d, %u, %u\n", motor100.velocity,
          motor100.current, motor100.temp);
  value += buf;
  sprintf(buf, "Motor220: %d, %u, %u\n", motor220.velocity,
          motor220.current, motor220.temp);
  value += buf;
  sprintf(buf, "Motor340: %d, %u, %u\n", motor340.velocity,
          motor340.current, motor340.temp);
  value += buf;
  sprintf(buf, "\nAccelerometer: %d %d %d\n", accel.x, accel.y, accel.z);
  value += buf;
  sprintf(buf, "Gyroscope: %d\n", rotation);
  value += buf;
  sprintf(buf, "Magnetometer: %d %d %d\n", mag.x, mag.y, mag.z);
  value += buf;
  sprintf(buf, "IR Distance: %u %u %u %u %u %u %u %u %u %u\n", ir_distance[0], 
    ir_distance[1], ir_distance[2], ir_distance[3], ir_distance[4], 
    ir_distance[5], ir_distance[6], ir_distance[7], ir_distance[8],
    ir_distance[9]);
  value += buf;
  sprintf(buf, "Light Sensors: %u %u %u %u\n", light[0], light[1],
      light[2], light[3]);
  value += buf;
  sprintf(buf, "\nError Codes: %08x\n", error_flags);
  value += buf;
  sprintf(buf, "CPU \%: %u\n", cycle_time);
  value += buf;


  return offset + value;
}

File: old-simulator/CritterStateDrop.java
/**
  * CritterStateDrop
  *
  * This class defines the various bits of subjective data that the robot
  *  has access to. Simulator-wise, this is what we provide to agents. 
  *  It should be a copy of CritterStateDrop.{h,cpp} from the C++ side.
  */

import java.io.IOException;
import java.util.EnumSet;

public class CritterStateDrop implements SimulatorDrop
{
  class motor_struct
  {
    public int velocity;
    public int current;
    public int temp;

    public void readData(InterfaceInputStream pIn) throws IOException
    {
      velocity = pIn.readInt();
      current = pIn.readInt();
      temp = pIn.readInt();
    }
    
    public void writeData(InterfaceOutputStream pOut) throws IOException
    {
      pOut.writeInt(velocity);
      pOut.writeInt(current);
      pOut.writeInt(temp);
    }

    public int getSize()
    {
      return 3 * Integer.SIZE;
    }
  }

  class vector3d
  {
    public int x, y, z;

    public void readData(InterfaceInputStream pIn) throws IOException
    {
      x = pIn.readInt();
      y = pIn.readInt();
      z = pIn.readInt();
    }
    
    public void writeData(InterfaceOutputStream pOut) throws IOException
    {
      pOut.writeInt(x);
      pOut.writeInt(y);
      pOut.writeInt(z);
    }

    public int getSize()
    {
      return 3 * Integer.SIZE;
    }
  }

  public enum PowerSource { SHORE, BAT40, BAT160, BAT280 };

  public PowerSource power_source;

  public int bus_voltage;
  public int batv40, batv160, batv280;

  public motor_struct motor100, motor220, motor340;
  public vector3d accel, mag;

  public int rotation;
  public int[] ir_distance;
  public int[] ir_light;
  public int[] light;
  public int[] thermal;
  public int[] bump;

  public int error_flags;
  public int cycle_time;

  public final int IR_DISTANCE_SIZE   = 10;
  public final int IR_LIGHT_SIZE      = 8;
  public final int LIGHT_SIZE         = 4;
  public final int THERMAL_SIZE       = 8;
  public final int BUMP_SIZE          = 32;

  /** Returns the size of the data contained in this Drop.
    * This value must correspond to the amount of data sent/received in
    *  writeData/readData.
    *
    * @return The size of the data contained in this Drop
    */
  public int getSize()
  {
    return (Integer.SIZE + // power_source (as an int)
           Integer.SIZE + // bus_voltage
           3 * Integer.SIZE + // batv40 + batv160 + batv280
           motor100.getSize() +
           motor220.getSize() + 
           motor340.getSize() +
           accel.getSize() +
           mag.getSize() +
           Integer.SIZE + // rotation
           IR_DISTANCE_SIZE * Integer.SIZE +
           IR_LIGHT_SIZE * Integer.SIZE +
           LIGHT_SIZE * Integer.SIZE +
           THERMAL_SIZE * Integer.SIZE +
           BUMP_SIZE * Integer.SIZE +
           Integer.SIZE + // error_flags
           Integer.SIZE) // cycle_time
           / 8; // divide by 8 because all of these are bit sizes
  }

  public CritterStateDrop()
  {
    power_source = PowerSource.SHORE;
    
    ir_distance = new int[IR_DISTANCE_SIZE];
    ir_light    = new int[IR_LIGHT_SIZE];
    light       = new int[LIGHT_SIZE];
    thermal     = new int[THERMAL_SIZE];
    bump        = new int[BUMP_SIZE];

    accel = new vector3d();
    mag = new vector3d();
    motor100 = new motor_struct();
    motor220 = new motor_struct();
    motor340 = new motor_struct();
  }

  public void writeData(InterfaceOutputStream pOut) throws IOException
  {
    pOut.writeInt(power_source.ordinal());

    pOut.writeInt(bus_voltage);
    pOut.writeInt(batv40);
    pOut.writeInt(batv160);
    pOut.writeInt(batv280);

    motor100.writeData(pOut);
    motor220.writeData(pOut);
    motor340.writeData(pOut);
    accel.writeData(pOut);
    mag.writeData(pOut);

    pOut.writeInt(rotation);
    for (int i = 0; i < ir_distance.length; i++)
      pOut.writeInt(ir_distance[i]);
    for (int i = 0; i < ir_light.length; i++)
      pOut.writeInt(ir_light[i]);
    for (int i = 0; i < light.length; i++)
      pOut.writeInt(light[i]);
    for (int i = 0; i < thermal.length; i++)
      pOut.writeInt(thermal[i]);
    for (int i = 0; i < bump.length; i++)
      pOut.writeInt(bump[i]);

    pOut.writeInt(error_flags);
    pOut.writeInt(cycle_time);
  }

  public void readData(InterfaceInputStream pIn) throws IOException
  {
    power_source = (PowerSource)EnumSet.range(PowerSource.SHORE,
      PowerSource.BAT280).toArray()[pIn.readInt()];
    
    // Read in all the data, one variable at a time
    bus_voltage = pIn.readInt();
    batv40 = pIn.readInt();
    batv160 = pIn.readInt();
    batv280 = pIn.readInt();

    motor100.readData(pIn);
    motor220.readData(pIn);
    motor340.readData(pIn);
    accel.readData(pIn);
    mag.readData(pIn);

    rotation = pIn.readInt();
    for (int i = 0; i < ir_distance.length; i++)
      ir_distance[i] = pIn.readInt();
    for (int i = 0; i < ir_light.length; i++)
      ir_light[i] = pIn.readInt();
    for (int i = 0; i < light.length; i++)
      light[i] = pIn.readInt();
    for (int i = 0; i < thermal.length; i++)
      thermal[i] = pIn.readInt();
    for (int i = 0; i < bump.length; i++)
      bump[i] = pIn.readInt();

    error_flags = pIn.readInt();
    cycle_time = pIn.readInt();
  }
}
